openapi3/swagger_utils_test.go

```diff
@@ -0,0 +1,33 @@
+package openapi3_test
+
+import (
+	"net/http"
+
+	"net/url"
+	"testing"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/stretchr/testify/require"
+)
+
+func TestResettingExternalRefs(t *testing.T) {
+
+	cs := startTestServer(http.Dir("testdata"))
+	defer cs()
+
+	loader := openapi3.NewSwaggerLoader()
+	loader.IsExternalRefsAllowed = true
+	remote, err := url.Parse("http://" + addr + "/test.refcache.openapi.yml")
+	require.NoError(t, err)
+
+	doc, err := loader.LoadSwaggerFromURI(remote)
+	require.NoError(t, err)
+
+	openapi3.ResetResolvedExternalRefs(doc)
```

Comment:
Did you explore combining this with a custom `loader.LoadSwaggerFromURIFunc`?

So we could do something like:
```
loader := openapi3.NewSwaggerLoader()
loader.LoadSwaggerFromURIFunc = MyLoaderThatResetsExtRefs
doc, err := loader.LoadSwaggerFromURI
```

Same number of steps, but maybe a new NewSwaggerLoader that accepts `...options`. Might not be worth it but feels more natural to configure the loader and then just load without an additional fixup step.

Reply:
Inelegant, but we could provide a Load function that creates a new loader (without a LoadSwaggerFromURIFunc set), or just temporarily nuke that field. Which then calls LoadSwaggerFromURIFunc which does all the resolution and loading. Then call Reset refs and returns from the original LoadSwaggerFromURI.

Just a thought.

Reply:
Keep in mind that oapi-codegen has conflicting requirements from the same swagger. 
Default code generation requires the references to be there since they get resolved to type names.
The spec generation required for validation cannot use these references, rather just the values.
Since typical calls to the generator make use of multiple options, I wanted to be able to perform this on the swagger that was parsed once, rather than have to perform a specific spec generation.

Having said that, it occurs to me that perhaps we can add another mode to the generator (`validation`) which then makes use of the approach you're suggesting.

We may also want to balance this with the fact that every time we call kin-openapi, there's a lot of network traffic that takes place. Separating the codegen calls will cause multiple such trips.

Happy to go either way

Reply:
I like the idea of having a validation mode, but noted on the network calls.

This would mean we use oapi-codegen to build golang code, causing N round trips. Then there is a go:generate step or gitlab-ci step which does the same thing but to create a validating specification which is the same N trips.

This doesn't happen every build though, it should be short-circuited the majority of the time because specs don't change as often as code - is my assumption of short-circuits correct?

Reply:
Indeed - this would only take place when processing the schema which should happen infrequently.

So the calls would be along the lines of:
`//go:generate <docker run> -p service -t server,types,client api.yaml > api.gen.go` followed by
`//go:generate <docker run> -p service -t validation api.yaml > api_validation.gen.go` 

I think I am warming up to this mode in oapi-codegen which means we can bundle that as an option to kin-openapi Load functions